app/jobs/hesa/retrieve_collection_job.rb

```diff
@@ -5,34 +5,36 @@ class RetrieveCollectionJob < ApplicationJob
     def perform
       return unless FeatureService.enabled?(:sync_from_hesa)
 
-      request_time = Time.zone.now
       xml_response = Hesa::Client.get(url: url)
 
-      # TODO: when we try to import the trainees we're going to mark
-      # the collection success/failure and include any failure in the subsequent run of this job
-      # This will prevent a failure here e.g. an error repsonse from incrementing the
-      # date and skipping a portion of the updates
+      Nokogiri::XML(xml_response).root.children.each do |student_node|
+        trainee, ukprn = Trainees::CreateFromHesa.call(student_node: student_node)
+        if trainee.invalid?
+          Sentry.capture_message("HESA import failed (errors: #{trainee.errors.full_messages}), (ukprn: #{ukprn})")
+          return save_hesa_request(xml_response).import_failed!
+        end
+      end
 
+      save_hesa_request(xml_response).import_successful!
+    end
+
+    def save_hesa_request(xml_response)
       HesaCollectionRequest.create(
-        requested_at: request_time,
+        requested_at: Time.zone.now,
```

Comment:
If we don't store the time before we make the request we'll run the risk of a long request causing a gap in the chain of request times. All assuming we can actually make requests for times rather than dates. Even then, if we run the thing before midnight and it takes until after midnight to complete I think this might skip a day.